### PR TITLE
Support NativeAOT and trimming in GeoIP2

### LINKED ISSUE #359: NativeAOT support
is it possible to upgrade that library to be NativeAOT compatible?

### LINKED ISSUE #505: Design Native AOT-compatible web-service JSON deserialization
# Design Native AOT-compatible web-service JSON deserialization

## Problem

Enabling the trimming and Native AOT analyzers for `MaxMind.GeoIP2` currently
reports `IL2026` and `IL3050` at the two reflection-based
`JsonSerializer.Deserialize<T>` calls in `WebServiceClient`.

Replacing those calls with a source-generated `JsonSerializerContext` appears
straightforward, but a prototype is not behavior-preserving with the current
public response models. The models are records with `init` properties, and
several properties use non-null default initializers. For JSON where a member is
missing, generated construction can pass `null` for that member and overwrite
the model's default initializer. In the current test suite, the direct
source-generation prototype caused nine web-service tests to fail, including
missing-key behavior and null-reference failures.

Annotating `WebServiceClient` as requiring dynamic code/unreferenced code is
also awkward. Database and web-service implementations share
`IGeoIP2Provider`; applying the attributes consistently to that interface would
make AOT-safe database methods appear unsafe to consumers, while applying them
only to the implementation produces analyzer contract warnings.

## Goal

Make the web-service path statically analyzable while preserving the existing
public model API and JSON semantics, especially defaults for missing members.
The package should only declare `IsAotCompatible` after both database and
web-service paths are analyzer-clean and covered by native-publish tests.

## Possible directions

1. Deserialize into internal mutable wire DTOs using a source-generated
   context, then map them to the public immutable response records. The wire
   DTOs can encode missing-member defaults explicitly.
2. Split or reshape the common provider abstraction so that any deliberately
   reflection-based web-service compatibility path can be annotated without
   contaminating the database API contract.
3. Explore custom source-generated converters/type-info customization that can
   distinguish missing JSON members from explicit `null`, if this can preserve
   all existing behavior without duplicating the model graph.

The internal DTO approach is likely the least surprising public API, although
it adds a second model graph and mapping maintenance.

## Suggested verification

- Keep all existing web-service tests unchanged and passing.
- Add explicit tests for missing members versus explicit `null` where defaults
  are non-null.
- Build with `IsAotCompatible=true` and
  `VerifyReferenceAotCompatibility=true`.
- Publish and execute a Native AOT smoke application that deserializes
  representative success and error responses.

The database-reader dependency has a related but separate model-mapping design
discussion in maxmind/MaxMind-DB-Reader-dotnet#321.

Would internal source-generated wire DTOs be an acceptable direction for a
follow-up implementation PR?

### PR DESCRIPTION
**Pull request by oschwald**
The database reader and web service client now support NativeAOT and trimming on .NET 8 and later. Built-in models use generated metadata without caller registration.

Closes [STF-1320](https://linear.app/maxmind/issue/STF-1320).
Closes #505.
Closes #359.

- Update MaxMind.Db to 5.2.0 and use an internal System.Text.Json context for successful responses and service errors.
- Preserve non-null model defaults when fields are omitted or explicitly null. This also makes direct null assignments use the property defaults.
- Replace stack inspection in database errors and avoid a native crash when cloning responses to apply locales.
- Enable compatibility analysis and run shared xUnit tests against the packed library under NativeAOT on .NET 9 and 10.
- Keep a small .NET 8 smoke test because the xUnit NativeAOT runner requires .NET 9 or later.
- Share web-service tests through injected HTTP handlers on modern .NET. Keep a local WireMock server for net481, whose synchronous transport bypasses those handlers. Only the two reflection-based JSON tests remain JIT-only. CI runs all three native targets on Linux, Windows, and macOS.

Validation:

- Release solution build: zero warnings and errors.
- Unit tests: 489 passed across .NET 8, 9, and 10.
- Package validation and Linux NativeAOT publish/run passed with SDK 10.0.401, full trimming, and JSON reflection disabled: 162 shared tests on .NET 9, 162 on .NET 10, and the .NET 8 smoke test.
- The script publishes and runs native binaries and requires at least 162 tests per run. Runtime assertions check that dynamic code and default JSON reflection are disabled.
- Markdown/YAML formatting and shell syntax checks passed.
- The net481 test project and the rewritten HTTP-fixture commit build with zero warnings and errors.
- CI passed on Linux, Windows, and macOS, including all three NativeAOT targets.
- Windows and the documentation CI build passed all 652 tests across net481 and .NET 8, 9, and 10.

## Summary by CodeRabbit

* **New Features**
  * Added NativeAOT and trimming support for the database reader and web service client on .NET 8 and later.
  * Added built-in source-generated serialization support without requiring response model registration.

* **Bug Fixes**
  * Preserved model defaults and collections when response fields are missing or null.
  * Database lookup errors now identify the public lookup method.
  * Improved handling and reporting of invalid or truncated database data.

* **Documentation**
  * Added guidance for enabling NativeAOT and trimming and running integration tests.

### CODE REVIEW

File: MaxMind.GeoIP2.UnitTests/ResponseTests.cs [hunk 1]
```diff
@@ -1,11 +1,60 @@
 using MaxMind.GeoIP2.Model;
 using MaxMind.GeoIP2.Responses;
+using System.Collections.Generic;
 using Xunit;
 
 namespace MaxMind.GeoIP2.UnitTests
 {
     public class ResponseTests
     {
+        public static TheoryData<AbstractCountryResponse> LocaleResponses => new()
```
[the remark this answers is not in the document]
Comment (oschwald) [reply MaxMind.GeoIP2.UnitTests/ResponseTests.cs:10]: Agent responding on behalf of Greg: Here, “public” means the library’s public API, not test methods declared public for the test runner. Documentation on test methods adds little value unless there is genuine uncertainty about what a test covers. In that case, a short explanatory comment is enough. XML documentation would be excessive. We will leave these test declarations without XML documentation.

File: MaxMind.GeoIP2.UnitTests.NativeAot/RuntimeTests.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
+using System.Runtime.CompilerServices;
+using System.Text.Json;
+using Xunit;
+
+namespace MaxMind.GeoIP2.UnitTests
+{
+    public class RuntimeTests
+    {
+        [Fact]
+        public void RunsAsNativeAotWithoutJsonReflection()
```
[the remark this answers is not in the document]
Comment (oschwald) [reply MaxMind.GeoIP2.UnitTests.NativeAot/RuntimeTests.cs]: Agent responding on behalf of Greg: Here, “public” means the library’s public API, not test methods declared public for the test runner. Documentation on test methods adds little value unless there is genuine uncertainty about what a test covers. In that case, a short explanatory comment is enough. XML documentation would be excessive. We will leave these test declarations without XML documentation.

File: README.dev.md [hunk 3]
```diff
@@ -3,6 +3,65 @@
...
+The script requires the .NET 10 SDK and the platform's
+[NativeAOT prerequisites](https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/#prerequisites).
+
+Both native test projects stay outside the solution. Ordinary restore and build
+use a reference to the library project. Publishing requires the package version
+supplied by the script so the native tests validate the packed library.
+
+## JSON defaults workaround
+
+The null-coalescing `init` accessors in the models and responses preserve
```
Comment (oschwald) [inline README.dev.md:43]: I created a follow-up issue to clean this up once `System.Text.Json` 11 is released.

File: MaxMind.GeoIP2.NativeAot/Program.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
+using MaxMind.GeoIP2;
+using System;
+using System.Net;
+using System.Net.Http;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+// xUnit's NativeAOT runner requires .NET 9. Keep a small .NET 8 smoke test.
+if (RuntimeFeature.IsDynamicCodeSupported || System.Text.Json.JsonSerializer.IsReflectionEnabledByDefault)
```
Comment (horgh) [inline MaxMind.GeoIP2.NativeAot/Program.cs:10]: This guard cannot detect the condition its message describes, so the smoke test also passes under the normal managed runtime.

`PublishAot=true` writes both feature switches into `runtimeconfig.json` at build time, so a plain `dotnet run -f net8.0` of this project passes the guard and prints ".NET 8 NativeAOT smoke test passed." without a native executable being involved. I confirmed the same behavior on the sibling test project: the managed `bin/Debug/net10.0/MaxMind.GeoIP2.UnitTests.dll` ran under the normal runtime and passed 162/162.

The script always runs the published native binary, so this is not a hole in the strategy today. But the guard gives more confidence than it earns if the app is ever run directly.

🤖 Comment by Claude (Claude Code) on behalf of Will.

File: MaxMind.GeoIP2.UnitTests.NativeAot/RuntimeTests.cs (hunk 2 above)
Comment (horgh) [inline MaxMind.GeoIP2.UnitTests.NativeAot/RuntimeTests.cs]: The test name claims more than the assertions check. Both feature switches are set at build time, so this test also passes when the managed build of the assembly runs under the normal runtime.

`README.dev.md` already says the switches "do not prove native execution", so this is naming drift rather than a coverage hole. A name such as `JsonReflectionAndDynamicCodeAreDisabled` would match what is asserted.

🤖 Comment by Claude (Claude Code) on behalf of Will.

File: dev-bin/test-native-aot.sh [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+# Use `pwd -W` under Git Bash so MAXMIND_TEST_BASE_DIR reaches the native test
+# binary as a Windows path. MSYS converts command arguments, not environment
+# variables. `pwd -W` fails on other shells, which then fall back to `pwd`.
+repository_root="$(cd "$(dirname "${BASH_SOURCE[0]}")/.." && { pwd -W 2>/dev/null || pwd; })"
+readonly repository_root
+readonly library_project="$repository_root/MaxMind.GeoIP2/MaxMind.GeoIP2.csproj"
+readonly app_project="$repository_root/MaxMind.GeoIP2.NativeAot/MaxMind.GeoIP2.NativeAot.csproj"
+
+# Strip the CR that the native Windows dotnet writes, as $(...) removes only LF.
+version_prefix="$(dotnet msbuild "$library_project" -getProperty:VersionPrefix -p:TargetFramework=net8.0 | tr -d '\r')"
```
Comment (horgh) [inline dev-bin/test-native-aot.sh:15]: `version_prefix` captures the whole stdout of `dotnet msbuild`, not just the property value.

On a clean machine `dotnet msbuild` writes a first-run banner ahead of the value, and an MSBuild warning would also be included. `package_version` then holds that extra text, and `dotnet pack` fails with a confusing invalid-version error instead of a clear one. Adding `-nologo` and piping through `tail -1` would make this robust.

🤖 Comment by Claude (Claude Code) on behalf of Will.

File: MaxMind.GeoIP2.UnitTests.NativeAot/MaxMind.GeoIP2.UnitTests.NativeAot.csproj [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    <Error Text="Set GeoIP2PackageVersion to the package under test, or run dev-bin/test-native-aot.sh." />
+  </Target>
+
+  <ItemGroup>
+    <PackageReference Include="xunit.v3.aot.mtp-v2" Version="4.0.0" />
+    <ProjectReference Include="../MaxMind.GeoIP2/MaxMind.GeoIP2.csproj"
+                      Condition="'$(GeoIP2PackageVersion)' == ''" />
+    <PackageReference Include="MaxMind.GeoIP2" Version="$(GeoIP2PackageVersion)"
+                      Condition="'$(GeoIP2PackageVersion)' != ''" />
+    <Compile Include="../MaxMind.GeoIP2.UnitTests/**/*.cs"
```
Comment (horgh) [inline MaxMind.GeoIP2.UnitTests.NativeAot/MaxMind.GeoIP2.UnitTests.NativeAot.csproj]: This glob also matches paths under `MaxMind.GeoIP2.UnitTests/TestData/MaxMind-DB`, a submodule this repository does not control.

The submodule holds no `.cs` files today (it is a Go repository), so nothing breaks now. If one is ever added, it is compiled into this project and can break the build for reasons unrelated to this repository. Adding `../MaxMind.GeoIP2.UnitTests/TestData/**` to the `Exclude` would prevent that.

🤖 Comment by Claude (Claude Code) on behalf of Will.

**Review by horgh — COMMENTED** [review]
LGTM. Claude has a few minor comments.

File: MaxMind.GeoIP2.NativeAot/Program.cs (hunk 4 above)
Comment (oschwald) [reply MaxMind.GeoIP2.NativeAot/Program.cs:10]: Agreed. In cb18414, I changed the guard message to describe the feature switches and point to dev-bin/test-native-aot.sh. I also removed the native-execution claims from the failure and success messages. The assertions are unchanged. The script still publishes and runs the native executable.

The Linux .NET 8 smoke test and all 162 NativeAOT tests on each of .NET 9 and .NET 10 passed.

Codex agent responding on behalf of Greg.

File: MaxMind.GeoIP2.UnitTests.NativeAot/RuntimeTests.cs (hunk 2 above)
Comment (oschwald) [reply MaxMind.GeoIP2.UnitTests.NativeAot/RuntimeTests.cs]: Renamed this to JsonReflectionAndDynamicCodeAreDisabled in 04a90e5. The assertions are unchanged. All 162 NativeAOT tests passed on each of .NET 9 and .NET 10 on Linux.

Codex agent responding on behalf of Greg.

File: dev-bin/test-native-aot.sh (hunk 5 above)
Comment (oschwald) [reply dev-bin/test-native-aot.sh:15]: I could not reproduce the first-run banner with this command. With a fresh DOTNET_CLI_HOME and first-run setup enabled, stdout was exactly `6.1.0\n` and stderr was empty. The [SDK explicitly skips first-run setup for -getProperty](https://github.com/dotnet/sdk/blob/v10.0.302/src/Cli/dotnet/Program.cs#L216-L231).

I left the command unchanged. I have not found a warning that contaminates this query, and `tail -1` would discard unexpected output without validating that the remaining line is a version. If we find a concrete warning case, we can handle it explicitly.

Codex agent responding on behalf of Greg.

**Review by horgh — APPROVED** [review]
